### PR TITLE
python reader en-zh

### PR DESCRIPTION
**Pull request by HongyingG**
python-reader1
python-reader2
python-reader3
python-reader4
python-reader5
虚拟机上浏览器的截屏出了些问题，截屏查看可能不太方便，见谅~

### CODE REVIEW

**Comment by shanyi15** [comment]
please change the file name to reader_cn, otherwise the English doc will be covered.

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 1]
```diff
@@ -3,27 +3,25 @@
 ```
 
 # Python Reader
+在训练期和测试期，PaddlePaddle程序需要读数据。为了帮助用户写能执行读数据的代码，我们做了如下定义：
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: 在模型训练和预测阶段，PaddlePaddle程序需要读取训练或预测数据。为了帮助用户编写数据读取的代码，我们提供了如下接口：

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 2]
```diff
@@ -3,27 +3,25 @@
...
 
-During the training and testing phases, PaddlePaddle programs need to read data. To help the users write code that performs reading input data, we define the following:
+- *reader*: 用于读数据的函数，数据来自文件、网络、随机数生成器等等，迭代数据项。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: - *reader*: 用于读取数据的函数，数据可来自于文件、网络、随机数生成器等，函数每次返回一个数据项。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 3]
```diff
@@ -3,27 +3,25 @@
...
+- *reader creator*: 返回reader函数的函数。
+- *reader decorator*: 一个函数，接受一个或多个reader，并返回一个reader。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: - *reader decorator*: 接受一个或多个reader作为参数、返回一个新reader的函数。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 4]
```diff
@@ -3,27 +3,25 @@
...
+- *batch reader*: 用于读取数据的函数，数据来自*reader*、文件、网络、随机数生成器等等，迭代一批数据项。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: - *batch reader*: 用于读取数据的函数，数据可来自于文件、网络、随机数生成器等，函数每次返回一个batch大小的数据项。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 5]
```diff
@@ -3,27 +3,25 @@
...
 
-- A *reader*: A function that reads data (from file, network, random number generator, etc) and yields the data items.
-- A *reader creator*: A function that returns a reader function.
-- A *reader decorator*: A function, which takes in one or more readers, and returns a reader.
-- A *batch reader*: A function that reads data (from *reader*, file, network, random number generator, etc) and yields a batch of data items.
+同样提供一个函数，可以将reader转换成batch reader，会频繁用到reader creator和reader decorator。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: 此外，还提供了将reader转换为batch reader的函数，会频繁用到reader creator和reader decorator。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 6]
```diff
@@ -3,27 +3,25 @@
...
 
-and also provide a function which can convert a reader to a batch reader, frequently used reader creators and reader decorators.
-
-## Data Reader Interface
-
-*Data reader* doesn't have to be a function that reads and yields data items. It can just be any function without any parameters that creates an iterable (anything can be used in `for x in iterable`) as follows:
+## Data Reader 接口
+*Data reader*不一定要求为读取和遍历数据项的函数。可以仅为任意不带参数的函数,该函数可以创建一个迭代器，通过`for x in iterable`访问任意数据项，迭代器定义如下：
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: *Data reader*不一定要求为读取和遍历数据项的函数。它可以是返回iterable对象（即可以用于`for x in iterable`的任意对象）的任意不带参数的函数：

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 7]
```diff
@@ -3,27 +3,25 @@
...
-*Data reader* doesn't have to be a function that reads and yields data items. It can just be any function without any parameters that creates an iterable (anything can be used in `for x in iterable`) as follows:
+## Data Reader 接口
+*Data reader*不一定要求为读取和遍历数据项的函数。可以仅为任意不带参数的函数,该函数可以创建一个迭代器，通过`for x in iterable`访问任意数据项，迭代器定义如下： 
 
 ```
 iterable = data_reader()
 ```
 
-The item produced from the iterable should be a **single** entry of data and **not** a mini batch. The entry of data could be a single item or a tuple of items. Item should be of one of the [supported types](http://www.paddlepaddle.org/doc/ui/data_provider/pydataprovider2.html?highlight=dense_vector#input-types) (e.g., numpy 1d array of float32, int, list of int etc.)
+迭代产生的数据项应为**单项**或者元组（tuple）形式的数据而**不是**mini batch,应该在[支持的类型](http://www.paddlepaddle.org/doc/ui/data_provider/pydataprovider2.html?highlight=dense_vector#input-types) 中，例如float32,int类型的numpy一维矩阵，int类型的列表等。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: Iterable对象应产生单项或tuple形式的数据，而不是一个mini batch的数据。产生的数据项应在[支持的类型](http://www.paddlepaddle.org/doc/ui/data_provider/pydataprovider2.html?highlight=dense_vector#input-types) 中，例如float32,int类型的numpy一维矩阵，int类型的列表等。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 8]
```diff
@@ -42,42 +41,41 @@ def reader_creator_random_image_and_label(width, height, label):
     return reader
 ```
 
-## Batch Reader Interface
+## Batch Reader 接口
+*Batch reader*可以为任意不带参数的函数,该函数可以创建一个迭代器，通过`for x in iterable`访问任意数据项。迭代器的输出应为一批（列表）数据项，列表里的每项应为元组（tuple）。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: *Batch reader*可以是返回iterable对象（即可以用于for x in iterable的任意对象）的任意不带参数的函数。Iterable的输出应为一个batch（list）的数据项。list中的每个数据项均为一个tuple元组。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 9]
```diff
@@ -90,13 +88,13 @@ def custom_batch_reader():
 mnist_random_image_batch_reader = custom_batch_reader
 ```
 
-## Usage
+## 使用
+以下是我们如何用PaddlePaddle的reader：
 
-Following is how we can use the reader with PaddlePaddle:
-The batch reader, a mapping from item(s) to data layer, the batch size and the number of total passes will be passed into `paddle.train` as follows:
+batch reader是从数据项到数据层的映射，批尺寸和总传递数将传给到 `paddle.train`：
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: batch reader是从数据项到数据层（data layer）的映射，batch size和总pass数通过以下方式传给 `paddle.train`：

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 10]
```diff
@@ -105,29 +103,26 @@ batch_reader = paddle.batch(paddle.dataset.mnist.train(), 128)
 paddle.train(batch_reader, {"image":0, "label":1}, 128, 10, ...)
 ```
 
-## Data Reader Decorator
-
-The *Data reader decorator* takes in a single reader or multiple data readers and returns a new data reader. It is similar to a [python decorator](https://wiki.python.org/moin/PythonDecorators), but it does not use `@` in the syntax.
+## Data Reader装饰器
+*Data reader decorator*接收一个单项data reader或者多项data reader，返回一个新的data reader。和[python decorator](https://wiki.python.org/moin/PythonDecorators)类似，但在语法上不使用`@`。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: ## Data Reader Decorator
*Data reader decorator*接收一个或多个reader对象作为参数，返回一个新的reader对象。它类似于[python decorator](https://wiki.python.org/moin/PythonDecorators)类似，但在语法上不需要写`@`。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 11]
```diff
@@ -105,29 +103,26 @@ batch_reader = paddle.batch(paddle.dataset.mnist.train(), 128)
...
 
-Since we have a strict interface for data readers (no parameters and return a single data item), a data reader can be used in a flexible way using data reader decorators. Following are a few examples:
+我们对data reader接口有严格限制（无参数并返回一个单数据项），一个data reader可以更灵活地运用，使用data reader装饰器。以下是一些示例：
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: 我们对data reader接口有严格限制（无参数并返回单个数据项），data reader可灵活地搭配data reader decorators使用。以下是一些示例：

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 12]
```diff
@@ -105,29 +103,26 @@ batch_reader = paddle.batch(paddle.dataset.mnist.train(), 128)
...
 
-### Prefetch Data
+### 预取回数据（缓存数据）
+由于读数据需要一些时间，而没有数据无法进行训练，总体来说预取（缓存）数据会是一个不错的办法。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: ### 数据预读取
由于读数据需要一些时间，而没有数据无法进行训练，因此一般而言数据预读取会是一个很好的方法。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 13]
```diff
@@ -105,29 +103,26 @@ batch_reader = paddle.batch(paddle.dataset.mnist.train(), 128)
...
+用`paddle.reader.buffered`预取回（缓存）数据：
 
 ```python
 buffered_reader = paddle.reader.buffered(paddle.dataset.mnist.train(), 100)
 ```
 
-`buffered_reader` will try to buffer (prefetch) `100` data entries.
-
-### Compose Multiple Data Readers
+`buffered_reader`将尝试预取回（缓存）`100`个数据项。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: `buffered_reader`将尝试缓存（预读取）`100`个数据项。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 14]
```diff
@@ -105,29 +103,26 @@ batch_reader = paddle.batch(paddle.dataset.mnist.train(), 128)
...
+我们对data reader接口有严格限制（无参数并返回一个单数据项），一个data reader可以更灵活地运用，使用data reader装饰器。以下是一些示例：
 
-### Prefetch Data
+### 预取回数据（缓存数据）
+由于读数据需要一些时间，而没有数据无法进行训练，总体来说预取（缓存）数据会是一个不错的办法。
 
-Since reading data may take some time and training can not proceed without data, it is generally a good idea to prefetch the data.
-
-Use `paddle.reader.buffered` to prefetch data:
+用`paddle.reader.buffered`预取回（缓存）数据：
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: 用`paddle.reader.buffered`预读取数据：

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 15]
```diff
@@ -105,29 +103,26 @@ batch_reader = paddle.batch(paddle.dataset.mnist.train(), 128)
...
 buffered_reader = paddle.reader.buffered(paddle.dataset.mnist.train(), 100)
 ```
 
-`buffered_reader` will try to buffer (prefetch) `100` data entries.
-
-### Compose Multiple Data Readers
+`buffered_reader`将尝试预取回（缓存）`100`个数据项。
 
-For example, if we want to use a source of real images (say reusing mnist dataset), and a source of random images as input for [Generative Adversarial Networks](https://arxiv.org/abs/1406.2661).
+### 组成多项Data Reader
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: ### 组合多个Data Reader

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 16]
```diff
@@ -146,38 +141,37 @@ true_reader = reader_creator_bool(True)
...
 ## Q & A
 
-### Why does a reader return only a single entry, and not a mini batch?
+### 为什么一个reader只返回单项而不是mini batch？
 
-Returning a single entry makes reusing existing data readers much easier (for example, if an existing reader returns 3 entries instead if a single entry, the training code will be more complicated because it need to handle cases like a batch size 2).
+返回单项，可以更容易地复用已有的data reader，例如如果一个已有的reader返回3项而不是一个单项，这样训练代码会更复杂，因为需要处理像批尺寸为2这样的例子。
 
-We provide a function: `paddle.batch` to turn (a single entry) reader into a batch reader.
+我们提供一个函数来将reader（一个单项）转换成一个batch reader。
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: 我们提供一个函数来将一个单项reader转换成一个batch reader。

File: doc/fluid/user_guides/howto/prepare_data/reader.md [hunk 17]
```diff
@@ -146,38 +141,37 @@ true_reader = reader_creator_bool(True)
...
 
-### Why do we need a batch reader, isn't is sufficient to give the reader and batch_size as arguments during training ?
+### 为什么需要一个batch raeder，在训练过程中给出reader和batch_size参数这样不足够吗？
 
-In most of the cases, it would be sufficient to give the reader and batch_size as arguments to the train method. However sometimes the user wants to customize the order of data entries inside a mini batch, or even change the batch size dynamically. For these cases using a batch reader is very efficient and helpful.
+在大多数情况下，在训练方法中给出reader和batch_size参数是足够的。但有时用户想自定义mini batch里数据项的顺序，或者动态改变批尺寸。在这些情况下用batch reader会非常高效有用。
 
-### Why use a dictionary instead of a list to provide mapping?
+### 为什么用字典而不是列表进行映射（map）？
```
Comment (sneaxiy) [inline doc/fluid/user_guides/howto/prepare_data/reader.md]: ### 为什么用字典而不是列表进行映射？

File: doc/fluid/user_guides/howto/prepare_data/reader.md (hunk 7 above)
Comment (sneaxiy) [reply doc/fluid/user_guides/howto/prepare_data/reader.md]: 另外，http://www.paddlepaddle.org/doc/ui/data_provider/pydataprovider2.html?highlight=dense_vector#input-types
这个链接不可用，麻烦更正。

**Comment by HongyingG** [comment]
Thank you for your review~modifications have been done according to the suggestions.

**Review by haowang101779990 — APPROVED** [review]
LGTM
